econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -0,0 +1,1056 @@
+                                               remainder='passthrough')
+            W_transformer = ColumnTransformer([('encode', OneHotEncoder(drop='first', sparse=False),
+                                                [ind for ind in categorical_inds
+                                                 if ind != feat_ind and ind not in hinds]),
+                                               ('drop', 'drop', hinds),
+                                               ('drop_feat', 'drop', feat_ind)],
+                                              remainder='passthrough')
+            # Use _ColumnTransformer instead of ColumnTransformer so we can get feature names
+            X_transformer = _ColumnTransformer([ind for ind in categorical_inds
+                                                if ind != feat_ind and ind in hinds],
+                                               [ind for ind in hinds
+                                                if ind != feat_ind and ind not in categorical_inds])
+
+            # Controls are all other columns of X
+            WX = WX_transformer.fit_transform(X)
+            # can't use X[:, feat_ind] when X is a DataFrame
+            T = _safe_indexing(X, feat_ind, axis=1)
+
+            W = W_transformer.fit_transform(X)
+            X_xf = X_transformer.fit_transform(X)
+            if W.shape[1] == 0:
+                # array checking routines don't accept 0-width arrays
+                W = None
+
+            # perform model selection
+            model_t = (_first_stage_clf(WX, T, automl=self.nuisance_models == 'automl')
+                       if discrete_treatment else _first_stage_reg(WX, T, automl=self.nuisance_models == 'automl'))
+
+            # For the global model, use a constant featurizer to fit the ATE
+            # Ideally, this could be PolynomialFeatures(degree=0, include_bias=True), but degree=0 is unsupported
+            # So instead we'll use our own class
+            featurizer = _ConstantFeatures()
+
+            # TODO: support other types of heterogeneity via an initializer arg
+            #       e.g. 'forest' -> ForestDML
+            if self.heterogeneity_model == 'linear':
+                est = LinearDML(model_y=self._model_y,
+                                model_t=model_t,
+                                featurizer=featurizer,
+                                discrete_treatment=discrete_treatment,
+                                fit_cate_intercept=False,
+                                linear_first_stages=False,
+                                random_state=123)
+            else:
+                est = CausalForestDML(model_y=self._model_y,
+                                      model_t=model_t,
+                                      featurizer=featurizer,
+                                      discrete_treatment=discrete_treatment,
+                                      random_state=123)
+                est.tune(y, T, X=X_xf, W=W)
```

Comment:
Tune doesn't seem to have been fixed and so this will through an error when the inputs are pandas dataframes